### PR TITLE
Refactor table schemas to catalog entry

### PR DESCRIPTION
**Pull request by acquamarin**
This is the first set of PR to change our catalog to use the catalogEntry based design.
This PR introduces the catalogEntry class and refactors the tableSchema to use the catalogEntry.
Tables names are case-insensitive now meaning that person and PeRson are the same table in kuzu.

### CODE REVIEW

File: src/binder/binder.cpp [hunk 1]
```diff
@@ -169,8 +169,9 @@ void Binder::validateReadNotFollowUpdate(const NormalizedSingleQuery& singleQuer
 }
 
 void Binder::validateTableType(table_id_t tableID, TableType expectedTableType) {
-    if (catalog.getTableSchema(clientContext->getTx(), tableID)->tableType != expectedTableType) {
-        throw BinderException("aa");
+    if (catalog.getTableCatalogEntry(clientContext->getTx(), tableID)->getTableType() !=
```
Comment (andyfengHKU) [inline src/binder/binder.cpp:172]: Let's add a test to cover this.

```
CREATE NODE TABLE N (id SERIAL, PRIMARY KEY(id));
CREATE REL TABLE R (FROM N TO N);
CREATE REL TABLE R2 (FROM R TO R);
```

File: src/catalog/catalog.cpp [hunk 2]
```diff
@@ -22,19 +24,19 @@ Catalog::Catalog(WAL* wal, VirtualFileSystem* vfs) : isUpdated{false}, wal{wal}
...
 }
 
 bool Catalog::containsRelTable(Transaction* tx) const {
-    return getVersion(tx)->containsTable(TableType::REL);
+    return getVersion(tx)->containsTable(CatalogEntryType::REL_TABLE_ENTRY);
 }
 
 bool Catalog::containsRdfGraph(Transaction* tx) const {
-    return getVersion(tx)->containsTable(TableType::RDF);
+    return getVersion(tx)->containsTable(CatalogEntryType::RDF_GRAPH_ENTRY);
```
Comment (andyfengHKU) [inline src/catalog/catalog.cpp:39]: Remove if we are not using this.

File: src/catalog/catalog.cpp [hunk 3]
```diff
@@ -265,7 +282,8 @@ std::vector<std::string> Catalog::getMacroNames(transaction::Transaction* tx) co
 void Catalog::setTableComment(table_id_t tableID, const std::string& comment) {
     KU_ASSERT(readWriteVersion != nullptr);
     setToUpdated();
-    readWriteVersion->getTableSchema(tableID)->setComment(comment);
+    auto tableEntry = readWriteVersion->getTableCatalogEntry(tableID);
+    ku_dynamic_cast<CatalogEntry*, TableCatalogEntry*>(tableEntry)->setComment(comment);
```
Comment (andyfengHKU) [inline src/catalog/catalog.cpp:286]: I think can associate comment to catalog entry which means we can comment any catalog entry.

File: src/catalog/catalog_set.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+void CatalogSet::removeEntry(const std::string& name) {
+    KU_ASSERT(containsEntry(name));
+    entries.erase(std::move(name));
+}
+
+void CatalogSet::renameEntry(const std::string& oldName, const std::string& newName) {
+    KU_ASSERT(containsEntry(oldName));
+    if (oldName == newName) {
+        return;
```
Comment (andyfengHKU) [inline src/catalog/catalog_set.cpp:35]: Ditto coverage.

File: src/include/catalog/catalog_content.h [hunk 5]
```diff
@@ -50,38 +50,47 @@ class CatalogContent {
...
-    // TODO(Guodong): I don't think it's necessary to keep separate maps for node and rel tables.
-    std::unordered_map<common::table_id_t, std::unique_ptr<TableSchema>> tableSchemas;
     // These two maps are maintained as caches. They are not serialized to the catalog file, but
     // is re-constructed when reading from the catalog file.
     std::unordered_map<std::string, common::table_id_t> tableNameToIDMap;
     common::table_id_t nextTableID;
     std::unique_ptr<function::BuiltInFunctions> builtInFunctions;
     std::unordered_map<std::string, std::unique_ptr<function::ScalarMacroFunction>> macros;
     common::VirtualFileSystem* vfs;
+    std::unique_ptr<CatalogSet> tables;
```
Comment (andyfengHKU) [inline src/include/catalog/catalog_content.h:93]: If we introduce CatalogSet here (btw you probably don't need unique_ptr), I have following questions

- should we remove `std::unordered_map<std::string, common::table_id_t> tableNameToIDMap;` and `nextTableID` ?
- Do we have plan to make `macros` and `builtInFunctions` as part of CatalogSet

File: src/include/function/table_functions/call_functions.h [hunk 6]
```diff
@@ -200,7 +202,7 @@ struct StorageInfoBindData final : public CallTableFuncBindData {
             schema, this->columnTypes, this->columnNames, table);
     }
 
-    catalog::TableSchema* schema;
+    catalog::TableCatalogEntry* schema;
```
Comment (andyfengHKU) [inline src/include/function/table_functions/call_functions.h:205]: don't forget to change variable name across this file.

File: src/processor/operator/aggregate/aggregate_hash_table.cpp [hunk 7]
```diff
@@ -685,7 +685,7 @@ bool compareEntry<struct_entry_t>(
     return true;
 }
 
-static bool compareNodeEntry(
+static bool comparenodeTableEntry(
```
Comment (andyfengHKU) [inline src/processor/operator/aggregate/aggregate_hash_table.cpp:688]: revert this file

File: src/storage/wal_replayer.cpp [hunk 8]
```diff
@@ -252,7 +252,7 @@ void WALReplayer::replayDropTableRecord(const kuzu::storage::WALRecord& walRecor
     if (isCheckpoint) {
         auto tableID = walRecord.dropTableRecord.tableID;
         if (!isRecovering) {
-            auto tableSchema = catalog->getTableSchema(&DUMMY_READ_TRANSACTION, tableID);
+            auto tableSchema = catalog->getTableCatalogEntry(&DUMMY_READ_TRANSACTION, tableID);
```
Comment (andyfengHKU) [inline src/storage/wal_replayer.cpp:255]: change variable name across this file

File: tools/shell/embedded_shell.cpp [hunk 9]
```diff
@@ -11,10 +11,12 @@
 #include <sstream>
 
 #include "catalog/catalog.h"
+#include "catalog/catalog_entry/node_table_catalog_entry.h"
+#include "catalog/catalog_entry/rel_table_catalog_entry.h"
```
Comment (andyfengHKU) [inline tools/shell/embedded_shell.cpp:15]: I think we don't need these two includes.

**Review by andyfengHKU — APPROVED** [review]

File: src/include/binder/binder.h [hunk 10]
```diff
@@ -130,15 +131,16 @@ class Binder {
     /*** bind copy ***/
     std::unique_ptr<BoundStatement> bindCopyFromClause(const parser::Statement& statement);
     std::unique_ptr<BoundStatement> bindCopyNodeFrom(const parser::Statement& statement,
-        std::unique_ptr<common::ReaderConfig> config, catalog::TableSchema* tableSchema);
+        std::unique_ptr<common::ReaderConfig> config,
+        catalog::NodeTableCatalogEntry* nodeTableEntry);
     std::unique_ptr<BoundStatement> bindCopyRelFrom(const parser::Statement& statement,
-        std::unique_ptr<common::ReaderConfig> config, catalog::TableSchema* tableSchema);
+        std::unique_ptr<common::ReaderConfig> config, catalog::RelTableCatalogEntry* relTableEntry);
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/binder/binder.h:137]: We don't have to include "catalog/catalog_entry/rdf_graph_catalog_entry.h" in binder.h
The binder includes the catalog.h which already declared the RelTableCatalogEntry

File: src/include/catalog/catalog_content.h (hunk 5 above)
Comment (acquamarin) [reply src/include/catalog/catalog_content.h:93]: macros and builtInFunctions will have their own catalogSet. 
The schema catalogEntry would contain the following fields (copied from duckdb):
```
class DuckSchemaEntry : public SchemaCatalogEntry {
public:
	DuckSchemaEntry(Catalog &catalog, string name, bool is_internal);

private:
	//! The catalog set holding the tables
	CatalogSet tables;
	//! The catalog set holding the indexes
	CatalogSet indexes;
	//! The catalog set holding the table functions
	CatalogSet table_functions;
	//! The catalog set holding the copy functions
	CatalogSet copy_functions;
	//! The catalog set holding the pragma functions
	CatalogSet pragma_functions;
	//! The catalog set holding the scalar and aggregate functions
	CatalogSet functions;
	//! The catalog set holding the sequences
	CatalogSet sequences;
	//! The catalog set holding the collations
	CatalogSet collations;
	//! The catalog set holding the types
	CatalogSet types;
}
```

File: tools/shell/embedded_shell.cpp (hunk 9 above)
Comment (acquamarin) [reply tools/shell/embedded_shell.cpp:15]: I think we do, getNodeTables/getRelTables returns a vector<NodeTableCatalogEntry*> and vector<RelTableCatalogEntry*>. The shell needs to know the types.

**Pull request merged**